File: CharAnalysis_2_0_MATLAB/src/CharPeakID.m
function [Charcoal, CharThresh] = CharPeakID(Charcoal, Pretreatment, ...
        PeakAnalysis, CharThresh)
% CharPeakID    Identify charcoal peaks based on threshold values.
%   [Charcoal, CharThresh] = CharPeakID(Charcoal, Pretreatment,
%       PeakAnalysis, CharThresh)
%
%   Identifies interpolated samples where Cpeak exceeds each threshold
%   value, removes consecutive exceedances so that only one sample per
%   event is retained, then screens the identified peaks against a
%   minimum-count criterion.
%
%   INPUTS
%     Charcoal     : struct containing peak, accI, countI, volI, ybpI
%     Pretreatment : struct  (yrInterp)
%     PeakAnalysis : struct  (threshType, threshValues, minCountP)
%     CharThresh   : struct  (possible, pos, neg, minCountP)
%
%   OUTPUTS  (fields added to input structs)
%     Charcoal.charPeaks       : [N x T] binary peak matrix (1 = peak)
%     Charcoal.charPeaksThresh : [N x T] threshold value at each peak
%     Charcoal.peaksTotal      : [1 x T] total peak count per threshold
%     Charcoal.threshFRI       : fire-return intervals per threshold
%     CharThresh.minCountP     : [N x T] p-values from minimum-count test
%
%   v2.0 changes vs v1.1
%     - Peak flagging: O(N x T) double for-loop replaced by a single
%       vectorised broadcast comparison.  Produces identical results.
%     - Consecutive-peak removal: two sequential O(N x T) for-loops
%       replaced by a diff-based column-wise operation.  Produces
%       identical results (see note on trailing-element behaviour below).
%     - charPeaksThresh construction: O(N x T) for-loop replaced by a
%       single broadcast multiply.
%     - Minimum-count analysis loop retained as-is.  It iterates over
%       identified peaks (~10-200 per record), not over all N x T samples,
%       so vectorising it would add complexity with negligible speed gain.
%
%   NOTE ON CONSECUTIVE-PEAK REMOVAL
%     The v1.1 algorithm (and v2.0) retains the LAST sample of each
%     consecutive run of above-threshold values (highest array index =
%     oldest age in cal yr BP).  For a run of flagged samples the loop
%     halved every element whose successor was also flagged, leaving the
%     trailing element as the only one still at value 2 after the pass.
%     The comment in v1.1 said "keep first peak as 2" but the implemented
%     behaviour was the reverse.  v2.0 preserves the implemented behaviour
%     exactly and documents it correctly.
%
%   CharAnalysis v2.0  -  Phase 1 modernisation

%% ── LOCAL VARIABLES ───────────────────────────────────────────────────────
r         = Pretreatment.yrInterp;
N         = length(Charcoal.peak);
alphaPeak = PeakAnalysis.minCountP;

%% ── BUILD THRESHOLD VALUE MATRIX ─────────────────────────────────────────
%
%   thresholdValues is [N x T]:
%     Global threshold  - constant columns (one value per column, same for
%                         all rows); built from CharThresh.possible.
%     Local threshold   - CharThresh.pos is already [N x T].

if PeakAnalysis.threshType == 1

    % Global: restrict to the positive-valued portion of the threshold
    % grid, then replicate into an [N x T] matrix.
    CharThreshPossible = CharThresh.possible(CharThresh.possible > 0);
    nThresholds        = length(CharThreshPossible);
    thresholdValues    = repmat(CharThreshPossible, N, 1);   % [N x T]

else

    % Local: threshold varies by sample; [N x T] already.
    thresholdValues = CharThresh.pos;
    nThresholds     = size(thresholdValues, 2);

end

%% ── VECTORISED PEAK FLAGGING ─────────────────────────────────────────────
%
%   v1.1 used a double for-loop:
%       for i = 1:N
%           for j = 1:nThresholds
%               if Charcoal.peak(i) > thresholdValues(i,j)
%                   Charcoal.charPeaks(i,j) = 2;
%               ...
%
%   v2.0: Charcoal.peak is [N x 1]; thresholdValues is [N x T].
%   MATLAB broadcasts the comparison across all T columns simultaneously.
%   Exceedances are marked as 2 (not 1) so the consecutive-removal step
%   can use value 1 as a "downgraded" marker distinct from "no peak" (0).

Charcoal.charPeaks = 2 * double(Charcoal.peak > thresholdValues);

%% ── VECTORISED CONSECUTIVE-PEAK REMOVAL ──────────────────────────────────
%
%   v1.1 walked forward through the array twice:
%     Pass 1: if sample i AND sample i+1 are both flagged, halve sample i
%             (2 -> 1).  The last sample of each run therefore stays at 2.
%     Pass 2: anything < 2 -> 0;  anything >= 2 -> 1.
%   Net result: the LAST sample of each consecutive run is kept as 1.
%
%   v2.0 uses diff to find run-end positions in a single pass per column:
%     diff(pk)(i) < 0  is true when pk(i) = 1 and pk(i+1) = 0,
%     i.e. sample i is the last flagged sample before a gap.
%     Appending pk(end) handles a run that reaches the final sample.

for j = 1:nThresholds
    pk    = Charcoal.charPeaks(:, j) > 0;          % logical [N x 1]
    isEnd = [diff(pk) < 0; pk(end)];               % true at run-end samples
    Charcoal.charPeaks(:, j) = double(isEnd);
end

%% ── charPeaksThresh: THRESHOLD VALUE AT EACH IDENTIFIED PEAK ─────────────
%
%   Used only for graphing.  Records the threshold value that each peak
%   was identified against, so downstream plot code can draw the appropriate
%   marker height.
%
%   v1.1 used an O(N x T) for-loop using thresholdValues(1,:) for all rows.
%   v2.0: single broadcast multiply.  thresholdValues(1,:) is [1 x T];
%   Charcoal.charPeaks is [N x T]; MATLAB broadcasts the row across N rows.

Charcoal.charPeaksThresh = Charcoal.charPeaks .* thresholdValues(1, :);

%% ── MINIMUM-COUNT ANALYSIS ───────────────────────────────────────────────
%
%   For each identified peak, tests whether the maximum charcoal count
%   in a window around the peak is statistically greater than the minimum
%   count using the Detre-White / Shuie-Bain test for unequal sample
%   volumes (Gavin 2006).  Peaks that do not pass are flagged in
%   CharThresh.minCountP and removed from Charcoal.charPeaks below.
%
%   This loop iterates over identified peaks (typically 10-200 per record),
%   not over all N samples, so it is not a performance bottleneck.

mcWindow = round(150 / r) * r;     % [yr] search window around each peak
d        = zeros(N, nThresholds);
CharThresh.minCountP = NaN(N, nThresholds);

for j = 1:nThresholds

    peakIndex = find(Charcoal.charPeaks(:, j));

    if length(peakIndex) <= 1
        continue        % need at least 2 peaks to define inter-peak windows
    end

    for i = 1:length(peakIndex)

        peakYr = Charcoal.ybpI(peakIndex);

        % Time window bounded by mcWindow on each side
        windowTime = [ max(Charcoal.ybpI(Charcoal.ybpI <= peakYr(i) + mcWindow)), ...
                       min(Charcoal.ybpI(Charcoal.ybpI >= peakYr(i) - mcWindow)) ];
        windowTime_in = [ find(Charcoal.ybpI == windowTime(1)), ...
                          find(Charcoal.ybpI == windowTime(2)) ];

        % Narrow window to adjacent peaks when they fall closer than mcWindow
        if i == 1
            windowPeak_in = [ find(Charcoal.ybpI == Charcoal.ybpI(peakIndex(i+1))), ...
                              find(Charcoal.ybpI == windowTime(2)) ];
        elseif i == length(peakIndex)
            windowPeak_in = [ find(Charcoal.ybpI == windowTime(1)), ...
                              find(Charcoal.ybpI == Charcoal.ybpI(peakIndex(i-1))) ];
        else
            windowPeak_in = [ find(Charcoal.ybpI == Charcoal.ybpI(peakIndex(i+1))), ...
                              find(Charcoal.ybpI == Charcoal.ybpI(peakIndex(i-1))) ];
        end

        if windowTime_in(1) > windowPeak_in(1)
            windowTime_in(1) = windowPeak_in(1);
        end
        if windowTime_in(2) < windowPeak_in(2)
            windowTime_in(2) = windowPeak_in(2);
        end

        windowSearch = windowTime_in;

        % Maximum count AFTER the peak (younger side, lower indices)
        countMax = round(max(Charcoal.countI(windowSearch(2) : peakIndex(i))));
        countMaxIn = windowSearch(2) - 1 + ...
            find(round(Charcoal.countI(windowSearch(2) : peakIndex(i))) ...
                 == countMax, 1);

        % Minimum count BEFORE the peak (older side, higher indices)
        countMin = round(min(Charcoal.countI(peakIndex(i) : windowSearch(1))));
        countMinIn = peakIndex(i) - 1 + ...
            find(round(Charcoal.countI(peakIndex(i) : windowSearch(1))) ...
                 == countMin, 1);

        volMax = Charcoal.volI(countMaxIn);
        volMin = Charcoal.volI(countMinIn);

        % Shuie-Bain (1982) expansion of Detre-White (1970) test statistic
        % for unequal sediment volumes (from Gavin 2006 / Charster)
        d(peakIndex(i), j) = ...
            ( abs(countMin - (countMin + countMax) * ...
                  (volMin / (volMin + volMax))) - 0.5 ) / ...
            sqrt( (countMin + countMax) * ...
                  (volMin  / (volMin + volMax)) * ...
                  (volMax  / (volMin + volMax)) );

        % p-value: t-distribution at 1e10 df is numerically the standard
        % normal (df -> inf limit); tcdf from Statistics Toolbox used here.
        CharThresh.minCountP(peakIndex(i), j) = 1 - tcdf(d(peakIndex(i), j), 1e10);

    end % each peak
end % each threshold

%% ── REMOVE PEAKS FAILING THE MINIMUM-COUNT CRITERION ────────────────────

for j = 1:nThresholds
    failIdx = find( Charcoal.charPeaks(:, j)    >  0 & ...
                    CharThresh.minCountP(:, j)   > alphaPeak );
    Charcoal.charPeaks(failIdx, j)       = 0;
    Charcoal.charPeaksThresh(failIdx, j) = 0;
end

%% ── FIRE-RETURN INTERVAL SENSITIVITY INDICES ────────────────────────────

for j = 1:nThresholds
    Charcoal.peaksTotal(j) = sum(Charcoal.charPeaks(:, j));

    inFRI = diff(Charcoal.ybpI(Charcoal.charPeaks(:, j) > 0));
    if ~isempty(inFRI)
        Charcoal.threshFRI(1:length(inFRI), j) = inFRI;
    end
end

end

File: CharAnalysis_2_0_R/R/char_peak_id.R
#' Identify charcoal peaks and apply minimum-count screening
#'
#' Mirrors \code{CharPeakID.m} from the MATLAB v2.0 codebase.  For each
#' threshold column, samples where C_peak exceeds the threshold are flagged,
#' consecutive exceedances are collapsed to a single event (keeping the
#' \strong{last} sample of each run, i.e. the oldest, matching the MATLAB
#' v1.1 / v2.0 convention), and each event is screened against a
#' minimum-count criterion using the Shuie-Bain (1982) extension of the
#' Detre-White (1970) test for unequal sediment volumes.
#'
#' @param charcoal     Named list containing:
#'   \describe{
#'     \item{peak}{C_peak series (\code{accI - accIS} or \code{accI / accIS}),
#'       length \eqn{N}.}
#'     \item{ybpI}{Age at each interpolated sample (cal yr BP), length \eqn{N}.}
#'     \item{accI}{Interpolated charcoal accumulation rate, length \eqn{N}.}
#'     \item{countI}{Interpolated charcoal count, length \eqn{N}.}
#'     \item{volI}{Interpolated sediment volume (cm^3), length \eqn{N}.}
#'   }
#' @param pretreatment Named list with element \code{yrInterp} (interpolation
#'   resolution, years).
#' @param peak_analysis Named list with elements \code{threshType} (1 = global,
#'   2 = local), \code{threshValues} (numeric vector, length \eqn{T}), and
#'   \code{minCountP} (alpha for minimum-count screen).
#' @param char_thresh  Named list returned by [char_thresh_global()] or
#'   [char_thresh_local()], containing \code{possible} (251-bin grid) and
#'   \code{pos} (\eqn{[N \times T]} threshold matrix).
#'
#' @return A named list with two components:
#'   \describe{
#'     \item{charcoal}{Input \code{charcoal} list augmented with:
#'       \itemize{
#'         \item \code{charPeaks}      -- \eqn{[N \times T]} numeric: 1 at peak
#'           samples, 0 elsewhere.
#'         \item \code{charPeaksThresh} -- \eqn{[N \times T]} numeric: threshold
#'           value at each identified peak, 0 elsewhere.
#'         \item \code{peaksTotal}     -- numeric vector length \eqn{T}: total
#'           peaks per threshold column.
#'         \item \code{threshFRI}      -- numeric matrix (\eqn{\leq N \times T}):
#'           fire-return intervals derived from peak ages per threshold column.
#'       }
#'     }
#'     \item{char_thresh}{Input \code{char_thresh} list augmented with
#'       \code{minCountP} -- \eqn{[N \times T]} matrix of Shuie-Bain p-values
#'       (NaN where not computed).}
#'   }
#'
#' @details
#'   ## Threshold value matrix
#'   For **global** thresholds (\code{threshType == 1}), \code{char_thresh$pos}
#'   is a constant-row \eqn{[N \times T]} matrix reused directly.  For
#'   **local** thresholds (\code{threshType == 2}), \code{char_thresh$pos} is
#'   already \eqn{[N \times T]} (per-sample values).
#'
#'   ## Consecutive-peak removal
#'   After flagging all exceedances, a diff-based pass retains only the
#'   \emph{last} sample of each consecutive run -- the oldest sample within a
#'   group of contiguous above-threshold values.  This matches the MATLAB v1.1
#'   algorithm (which the v2.0 comment documents correctly despite the v1.1
#'   comment being misleading).
#'
#'   ## Minimum-count test
#'   For each identified peak \eqn{i} in column \eqn{j}, a time window of
#'   \eqn{\pm 150} yr is constructed around the peak, then narrowed to the
#'   adjacent peaks when they fall within the window.  The test statistic is
#'   \deqn{
#'     d = \frac{|c_{\min} - (c_{\min}+c_{\max})\,v_{\min}/(v_{\min}+v_{\max})| - 0.5}
#'              {\sqrt{(c_{\min}+c_{\max})\,v_{\min}\,v_{\max}/(v_{\min}+v_{\max})^2}}
#'   }
#'   and the p-value is \eqn{1 - \Phi(d)} (standard normal CDF; equivalent to
#'   MATLAB's \code{1 - tcdf(d, 1e10)} because \eqn{t_{1\times10^{10}} \to z}).
#'   Peaks with \eqn{p > \alpha_{\text{peak}}} are removed.
#'
#' @seealso [char_thresh_local()], [char_thresh_global()], [CharAnalysis()]
char_peak_id <- function(charcoal, pretreatment, peak_analysis, char_thresh) {

  r          <- pretreatment$yrInterp
  N          <- length(charcoal$peak)
  alpha_peak <- peak_analysis$minCountP

  # ============================================================
  # BUILD THRESHOLD VALUE MATRIX  [N x T]
  # ============================================================
  # Global threshold: restrict possible grid to positive values, replicate.
  # Local threshold:  char_thresh$pos is already [N x T].

  if (peak_analysis$threshType == 1L) {
    pos_grid         <- char_thresh$possible[char_thresh$possible > 0]
    n_thresholds     <- length(pos_grid)
    threshold_values <- matrix(rep(pos_grid, each = N),
                               nrow = N, ncol = n_thresholds)
  } else {
    threshold_values <- char_thresh$pos
    n_thresholds     <- ncol(threshold_values)
  }

  # ============================================================
  # PEAK FLAGGING  (vectorised broadcast, matches MATLAB v2.0)
  # ============================================================
  # Exceedances are marked 2 so the consecutive-removal step can
  # distinguish "downgraded" (1) from "absent" (0).

  char_peaks <- 2 * (outer(charcoal$peak, rep(1, n_thresholds)) >
                       threshold_values) * 1.0

  # ============================================================
  # CONSECUTIVE-PEAK REMOVAL  (diff-based, column-wise)
  # ============================================================
  # Retains the LAST sample of each consecutive run of exceedances
  # (= the oldest sample within each event group), matching the
  # implemented behaviour of MATLAB CharPeakID.m v1.1 and v2.0.

  for (j in seq_len(n_thresholds)) {
    pk    <- char_peaks[, j] > 0          # logical length-N
    is_end <- c(diff(pk) < 0, pk[N])      # TRUE at run-end positions
    char_peaks[, j] <- as.numeric(is_end)
  }

  # ============================================================
  # charPeaksThresh: threshold value at each identified peak
  # ============================================================
  # Used downstream for plotting and peak magnitude.
  # MATLAB uses thresholdValues(1,:) -- the first row (same for global;
  # per-column constant for local since the threshold is per-sample but
  # the graph marker only needs a representative value).

  char_peaks_thresh <- sweep(char_peaks, 2, threshold_values[1L, ], `*`)

  # ============================================================
  # MINIMUM-COUNT ANALYSIS
  # ============================================================
  mc_window <- round(150 / r) * r        # search half-window in years

  d_mat           <- matrix(0,         nrow = N, ncol = n_thresholds)
  min_count_p_mat <- matrix(NA_real_,  nrow = N, ncol = n_thresholds)

  for (j in seq_len(n_thresholds)) {

    peak_index <- which(char_peaks[, j] > 0)

    if (length(peak_index) <= 1L) next    # need >= 2 peaks

    for (i in seq_along(peak_index)) {

      peak_yr_i <- charcoal$ybpI[peak_index[i]]

      # ---- Time-based window +/-mcWindow around peak ----------------
      # ybpI is ordered youngest->oldest (ascending yr BP values).
      # windowTime[1]: oldest boundary (max ybpI <= peakYr + mcWindow)
      # windowTime[2]: youngest boundary (min ybpI >= peakYr - mcWindow)
      wt1_candidates <- charcoal$ybpI[charcoal$ybpI <= peak_yr_i + mc_window]
      wt2_candidates <- charcoal$ybpI[charcoal$ybpI >= peak_yr_i - mc_window]
      if (length(wt1_candidates) == 0L || length(wt2_candidates) == 0L) next

      window_time_1 <- max(wt1_candidates)
      window_time_2 <- min(wt2_candidates)
      wt_in_1 <- which(charcoal$ybpI == window_time_1)[1L]   # older end index
      wt_in_2 <- which(charcoal$ybpI == window_time_2)[1L]   # younger end index

      # ---- Narrow window to adjacent peaks ------------------------
      peak_yrs_all <- charcoal$ybpI[peak_index]

      if (i == 1L) {
        # Youngest peak: older side bounded by next peak (i+1)
        wp_in_1 <- which(charcoal$ybpI == peak_yrs_all[i + 1L])[1L]
        wp_in_2 <- wt_in_2
      } else if (i == length(peak_index)) {
        # Oldest peak: younger side bounded by previous peak (i-1)
        wp_in_1 <- wt_in_1
        wp_in_2 <- which(charcoal$ybpI == peak_yrs_all[i - 1L])[1L]
      } else {
        wp_in_1 <- which(charcoal$ybpI == peak_yrs_all[i + 1L])[1L]
        wp_in_2 <- which(charcoal$ybpI == peak_yrs_all[i - 1L])[1L]
      }

      # Apply narrowing: use peak boundary when it falls within time window
      if (!is.na(wp_in_1) && wt_in_1 > wp_in_1) wt_in_1 <- wp_in_1
      if (!is.na(wp_in_2) && wt_in_2 < wp_in_2) wt_in_2 <- wp_in_2

      ws <- c(wt_in_1, wt_in_2)   # ws[1]=older index, ws[2]=younger index

      # ---- Count extremes -----------------------------------------
      # countMax: max count in [ws[2] : peakIndex] (younger side)
      seg_young <- charcoal$countI[ws[2L]:peak_index[i]]
      count_max    <- round(max(seg_young))
      count_max_in <- ws[2L] - 1L +
        which(round(charcoal$countI[ws[2L]:peak_index[i]]) == count_max)[1L]

      # countMin: min count in [peakIndex : ws[1]] (older side)
      seg_old  <- charcoal$countI[peak_index[i]:ws[1L]]
      count_min    <- round(min(seg_old))
      count_min_in <- peak_index[i] - 1L +
        which(round(charcoal$countI[peak_index[i]:ws[1L]]) == count_min)[1L]

      vol_max <- charcoal$volI[count_max_in]
      vol_min <- charcoal$volI[count_min_in]

      # ---- Shuie-Bain (1982) test statistic -----------------------
      # Detre-White / Shuie-Bain test for unequal sediment volumes
      # (Gavin 2006 / Charster).
      v_ratio <- vol_min / (vol_min + vol_max)
      d_val   <- (abs(count_min - (count_min + count_max) * v_ratio) - 0.5) /
                 sqrt((count_min + count_max) * v_ratio * (1 - v_ratio))

      d_mat[peak_index[i], j] <- d_val

      # p-value: t_{1e10} -> standard normal; pnorm(d) is equivalent.
      min_count_p_mat[peak_index[i], j] <- 1 - stats::pnorm(d_val)
    }
  }

  # ============================================================
  # REMOVE PEAKS FAILING MINIMUM-COUNT CRITERION
  # ============================================================

  for (j in seq_len(n_thresholds)) {
    fail_idx <- which(char_peaks[, j] > 0 &
                      !is.na(min_count_p_mat[, j]) &
                      min_count_p_mat[, j] > alpha_peak)
    char_peaks[fail_idx, j]        <- 0
    char_peaks_thresh[fail_idx, j] <- 0
  }

  # ============================================================
  # FIRE-RETURN INTERVALS AND PEAK TOTALS
  # ============================================================

  peaks_total <- numeric(n_thresholds)
  thresh_fri  <- matrix(NA_real_, nrow = N, ncol = n_thresholds)

  for (j in seq_len(n_thresholds)) {
    peaks_total[j] <- sum(char_peaks[, j])
    fri_yrs <- diff(charcoal$ybpI[char_peaks[, j] > 0])
    if (length(fri_yrs) > 0L) {
      thresh_fri[seq_along(fri_yrs), j] <- fri_yrs
    }
  }

  # ============================================================
  # PACK RESULTS
  # ============================================================

  charcoal$charPeaks       <- char_peaks
  charcoal$charPeaksThresh <- char_peaks_thresh
  charcoal$peaksTotal      <- peaks_total
  charcoal$threshFRI       <- thresh_fri

  char_thresh$minCountP    <- min_count_p_mat

  list(charcoal   = charcoal,
       char_thresh = char_thresh)
}
